### PR TITLE
keras sequence methods for generator

### PR DESCRIPTION
**Pull request by bw4sz**
The goal of this PR is to introduce the keras sequence method (https://keras.io/models/sequential/) to the default generator class.

Goals 

1. Improve parallelization and thread-safe generators. Keras docs state:

"The use of keras.utils.Sequence guarantees the ordering and guarantees the single use of every input per epoch when using use_multiprocessing=True."

2. Implemented a __getitem__ and __len__ method.

3. Remove threading and locks.

4. Move group shuffling to explicit on_epoch_end method

5. Added command line arguments for the fit_generator method.

I cloned the master, built a new conda env, and ran the generator and generator csv tests. 

*Potential Drawbacks*

A number of windows users have complained about multiprocessing in keras. I found that multiprocessing was only useful when

1. Generator performs alot of processing. This could happen when files are very large and reading is the bottleneck. Or when the user creates a custom generator method (as I've done in my fork).

2. When there is sufficient memory to fill the queue. 

My production environment:

I am running linux on a large multi-node cluster, feeding in > 20,000 images during training in batches of 6. I found a huge performance boosts from using the sequence method and multiprocessing, but the overhead is probably too much for the csv generator method in most cases.

### CODE REVIEW

**Comment by bw4sz** [comment]
I also had to change the test to the generator, since there isn't a next method. I assume the goal was just to produce a batch, so I chose the first batch 

```
            _, [_, labels_batch] = simple_generator.__getitem__(0)
```

**Comment by de-vri-es** [comment]
I don't have time to review now properly, but this has been on our internal wishlist for some time now. So thank you for the PR. We'll review this soon :)

**Comment by bw4sz** [comment]
It looks like i need to call group_images() before shuffling on init. Pushing commit now.

File: keras_retinanet/bin/train.py [hunk 1]
```diff
@@ -409,6 +409,11 @@ def csv_list(string):
     parser.add_argument('--config',           help='Path to a configuration parameters .ini file.')
     parser.add_argument('--weighted-average', help='Compute the mAP using the weighted average of precisions among classes.', action='store_true')
 
+    #Fit generator arguments
+    parser.add_argument('--multiprocessing', help='Fit generator mulitprocessing for n worker processes', action='store_true')
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Minor detail, but the help section are missing a period at the end.

File: keras_retinanet/bin/train.py [hunk 2]
```diff
@@ -409,6 +409,11 @@ def csv_list(string):
...
+    parser.add_argument('--workers', help='Number of multiprocessing workers', default=1)
+    parser.add_argument('--max_queue_size', help='Queue length for multiprocessing workers in fit generator', default=10)
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Use dashes instead of underscores here.

File: keras_retinanet/preprocessing/generator.py [hunk 3]
```diff
@@ -36,8 +35,7 @@
 )
 from ..utils.transform import transform_aabb
 
-
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: I didn't check, but I'm guessing removing this newline causes PEP8 to complain.

File: keras_retinanet/preprocessing/generator.py [hunk 4]
```diff
@@ -311,16 +314,15 @@ def compute_input_output(self, group):
...
-        # advance the group index
-        with self.lock:
-            if self.group_index == 0 and self.shuffle_groups:
-                # shuffle groups at start of epoch
-                random.shuffle(self.groups)
-            group = self.groups[self.group_index]
-            self.group_index = (self.group_index + 1) % len(self.groups)
-
-        return self.compute_input_output(group)
+    #Keras Sequence methods
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: I'm fine with removing this comment.

File: keras_retinanet/preprocessing/generator.py [hunk 5]
```diff
@@ -311,16 +314,15 @@ def compute_input_output(self, group):
...
+    def __len__(self):
+        """Number of batches for generator"""
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: Period missing.

File: keras_retinanet/preprocessing/generator.py [hunk 6]
```diff
@@ -311,16 +314,15 @@ def compute_input_output(self, group):
...
+        return len(self.groups)
+    
+    def __getitem__(self,index):
+        """
+        Keras sequence method for generating batches
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: Period missing.

File: keras_retinanet/preprocessing/generator.py [hunk 7]
```diff
@@ -311,16 +314,15 @@ def compute_input_output(self, group):
...
-            group = self.groups[self.group_index]
-            self.group_index = (self.group_index + 1) % len(self.groups)
-
-        return self.compute_input_output(group)
+    #Keras Sequence methods
+    def __len__(self):
+        """Number of batches for generator"""
+        return len(self.groups)
+    
+    def __getitem__(self,index):
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: Whitespace missing.

File: keras_retinanet/preprocessing/generator.py [hunk 8]
```diff
@@ -311,16 +314,15 @@ def compute_input_output(self, group):
...
+        """
+        Keras sequence method for generating batches
+        """
+        group = self.groups[index]        
+        inputs,targets=self.compute_input_output(group)
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: Whitespaces missing.

File: keras_retinanet/preprocessing/generator.py [hunk 9]
```diff
@@ -311,16 +314,15 @@ def compute_input_output(self, group):
...
+        return inputs,targets
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: Whitespace missing.

File: keras_retinanet/bin/train.py [hunk 10]
```diff
@@ -409,6 +409,11 @@ def csv_list(string):
     parser.add_argument('--config',           help='Path to a configuration parameters .ini file.')
     parser.add_argument('--weighted-average', help='Compute the mAP using the weighted average of precisions among classes.', action='store_true')
 
+    #Fit generator arguments
+    parser.add_argument('--multiprocessing', help='Fit generator mulitprocessing for n worker processes', action='store_true')
+    parser.add_argument('--workers', help='Number of multiprocessing workers', default=1)
```
Comment (de-vri-es) [inline keras_retinanet/bin/train.py]: I'd like to see `--multiprocessing` and `--workers` combined into a single argument. A value of 0 (the default) could mean no multiprocessing, other values would indicate the number of workers.

File: tests/preprocessing/test_generator.py [hunk 11]
```diff
@@ -262,7 +262,7 @@ def test_complete(self):
         simple_generator = SimpleGenerator(input_bboxes_group, input_labels_group, image=input_image, num_classes=6)
         # expect a UserWarning
         with pytest.warns(UserWarning):
-            _, [_, labels_batch] = simple_generator.next()
+            _, [_, labels_batch] = simple_generator.__getitem__(0)
```
Comment (de-vri-es) [inline tests/preprocessing/test_generator.py]: Shouldn't this just be `simple_generator[0]`?

**Comment by bw4sz** [comment]
sure, thanks for the style feedback, i'll get it done tomorrow.

File: tests/preprocessing/test_generator.py (hunk 11 above)
Comment (de-vri-es) [reply tests/preprocessing/test_generator.py]: The tests don't get that far yet because of the style checks. Might make sense for us to split that into two separate tests so the functional bits can be checked even if the style bit gives errors.

File: keras_retinanet/bin/train.py (hunk 10 above)
Comment (de-vri-es) [reply keras_retinanet/bin/train.py]: Looks good, thanks :)

**Comment by de-vri-es** [comment]
flake8 is still complaining about there:

```
_____ FLAKE8-check(ignoring E201 E202 E203 E221 E241 E251 E402 E501 W504) ______
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/bin/train.py:412:5: E265 block comment should start with '# '
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/bin/train.py:415:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/bin/train.py:485:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/bin/train.py:486:5: E265 block comment should start with '# '
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/bin/train.py:491:1: W293 blank line contains whitespace
_____ FLAKE8-check(ignoring E201 E202 E203 E221 E241 E251 E402 E501 W504) ______
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:83:9: E265 block comment should start with '# '
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:85:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:86:9: E265 block comment should start with '# '
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:86:35: W291 trailing whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:89:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:325:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:327:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:328:25: E231 missing whitespace after ','
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:332:35: W291 trailing whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:333:15: E231 missing whitespace after ','
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:333:23: E225 missing whitespace around operator
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:334:1: W293 blank line contains whitespace
/home/travis/build/fizyr/keras-retinanet/keras_retinanet/preprocessing/generator.py:335:22: E231 missing whitespace after ','
```

**Comment by bw4sz** [comment]
Maybe i'm running tests the wrong, but i checked locally for the first file.

```
(keras_retinanet) MacBook-Pro:keras-retinanet ben$ pytest tests/preprocessing/test_generator.py --flake8
=================================================== test session starts ===================================================
platform darwin -- Python 3.6.7, pytest-4.0.1, py-1.7.0, pluggy-0.8.0
rootdir: /Users/ben/Documents/keras-retinanet, inifile: setup.cfg
plugins: flake8-1.0.2
collected 6 items

tests/preprocessing/test_generator.py ......                                                                        [100%]

==================================================== warnings summary =====================================================
/Users/ben/miniconda3/envs/keras_retinanet/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
  /Users/ben/miniconda3/envs/keras_retinanet/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
    return _inspect.getargspec(target)
  /Users/ben/miniconda3/envs/keras_retinanet/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
    return _inspect.getargspec(target)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================== 6 passed, 2 warnings in 1.94 seconds ===========================================
(keras_retinanet) MacBook-Pro:keras-retinanet ben$
```

**Comment by de-vri-es** [comment]
You have to install flake8 and run the tests with `--flake8` if you want to see the same linting errors that travis generates. Alternatively you can look at the test output on travis itself by clicking the red cross icon (or yellow/green indicator if the test is running or succeeded).

**Comment by bw4sz** [comment]
i am running with --flake8, see previous message

```
 pytest tests/preprocessing/test_generator.py --flake8
```

i'll try to see what's going on.

**Comment by bw4sz** [comment]
Okay, those whitespaces were invisible to me, I think my IDE adds them. I ran 

```
autopep8 --select=W293 --in-place keras_retinanet/preprocessing/generator.py
```
following : https://stackoverflow.com/questions/28159539/how-to-fix-the-w293-blank-line-contains-whitespace

and it fixed them. 

those two files pass style tests locally

```
(keras_retinanet) MacBook-Pro:keras-retinanet ben$ pytest --flake8 keras_retinanet/preprocessing/generator.py keras_retinanet/bin/train.py
================================================================== test session starts ===================================================================
platform darwin -- Python 3.6.7, pytest-4.0.1, py-1.7.0, pluggy-0.8.0
rootdir: /Users/ben/Documents/keras-retinanet, inifile: setup.cfg
plugins: flake8-1.0.2
collected 2 items

keras_retinanet/preprocessing/generator.py s                                                                                                       [ 50%]
keras_retinanet/bin/train.py s                                                                                                                     [100%]

==================================================================== warnings summary ====================================================================
/Users/ben/miniconda3/envs/keras_retinanet/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75
  /Users/ben/miniconda3/envs/keras_retinanet/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
    return _inspect.getargspec(target)
  /Users/ben/miniconda3/envs/keras_retinanet/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:75: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
    return _inspect.getargspec(target)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================================= 2 skipped, 2 warnings in 1.25 seconds ==
```

File: keras_retinanet/preprocessing/generator.py [hunk 12]
```diff
@@ -81,11 +80,16 @@ def __init__(
...
-
+        # Define groups
         self.group_images()
 
+        # Shuffle when initializing
+        if self.shuffle_groups:
+            self.on_epoch_end()
+
+    def on_epoch_end(self):
+        random.shuffle(self.groups)
```
[the remark this answers is not in the document]
Comment (de-vri-es) [reply keras_retinanet/preprocessing/generator.py]: Almost certainly yes. There should be a check added. Do you want to do it?
